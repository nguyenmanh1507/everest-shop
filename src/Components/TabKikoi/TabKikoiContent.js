// @flow

import * as React from 'react'
import cx from 'classnames'

type Props = {
  children: React.Node,
  tabContentIndex?: number,
  tabActive?: number,
  currentTab?: ?number
}

class TabKikoiContent extends React.Component<Props> {
  render() {
    const { children, tabActive, tabContentIndex, currentTab } = this.props

    return (
      <div
        className={cx('tab__content', {
          active: currentTab
            ? currentTab === tabContentIndex
            : tabActive === tabContentIndex
        })}
        data-test={`tab-content-${tabContentIndex || ''}`}
      >
        {children}
      </div>
    )
  }
}

export default TabKikoiContent
