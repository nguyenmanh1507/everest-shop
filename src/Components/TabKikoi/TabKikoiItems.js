// @flow

import React, { Component } from 'react'
import cx from 'classnames'

type Props = {
  setCurrentTab: (tab: number) => void,
  items: Array<string>,
  currentTab: ?number,
  tabActive: number
}

class TabKikoiItem extends Component<Props> {
  handleNavItemClick = (event: SyntheticEvent<HTMLButtonElement>) => {
    const tab = parseInt(event.currentTarget.getAttribute('data-item'), 10)
    this.props.setCurrentTab(tab)
  }

  render() {
    const { items, currentTab, tabActive } = this.props

    return (
      <div className="tab__nav tab__nav--center">
        {items.map((item, index) => (
          <button
            key={`tab-${item}-${index + 1}`}
            className={cx('tab__nav-item', {
              active: currentTab
                ? currentTab === index + 1
                : tabActive === index + 1
            })}
            data-item={index + 1}
            data-test={`tab-item-${index + 1}`}
            onClick={this.handleNavItemClick}
          >
            {item}
          </button>
        ))}
      </div>
    )
  }
}

export default TabKikoiItem
