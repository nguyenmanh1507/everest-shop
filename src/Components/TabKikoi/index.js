// @flow

export { default } from './TabKikoiContainer'
