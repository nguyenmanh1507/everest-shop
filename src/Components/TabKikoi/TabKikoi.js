// @flow

import React, { Component, Children, cloneElement } from 'react'

import TabKikoiItems from './TabKikoiItems'
import TabKikoiContent from './TabKikoiContent'

type Props = {
  children: any,
  setCurrentTab: (tab: number) => void,
  items: Array<string>,
  currentTab: ?number,
  tabActive: number
}

class TabKikoi extends Component<Props> {
  contentNodeTotal = 0

  render() {
    const { children, setCurrentTab, items, currentTab, tabActive } = this.props

    let i = 1

    const childrenWithProps = Children.map(children, child => {
      this.contentNodeTotal++

      return cloneElement(child, {
        currentTab,
        tabActive,
        tabContentIndex: child.type === TabKikoiContent ? i++ : null
      })
    })

    return (
      <div className="tab">
        <TabKikoiItems
          setCurrentTab={setCurrentTab}
          items={items}
          currentTab={currentTab}
          tabActive={tabActive}
        />
        <div className="tab__body">{childrenWithProps}</div>
      </div>
    )
  }
}

export default TabKikoi
