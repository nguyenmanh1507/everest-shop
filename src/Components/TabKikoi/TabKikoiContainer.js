// @flow

import * as React from 'react'

import TabKikoi from './TabKikoi'
import TabKikoiContent from './TabKikoiContent'

type Props = {
  /** Tab content */
  children: React.Node,
  /** Tab items text */
  items: Array<string>,
  /** Set tab item active */
  tabActive: number
}

type State = {
  currentTab: ?number
}

class TabKikoiContainer extends React.Component<Props, State> {
  static displayName = 'TabKikoi'
  static Content = TabKikoiContent

  static defaultProps = {
    tabActive: 1
  }

  state = {
    currentTab: null
  }

  setCurrentTab = (tab: number) => {
    this.setState({
      currentTab: tab
    })
  }

  render() {
    const { currentTab } = this.state
    const { children, items, tabActive } = this.props

    return (
      <TabKikoi
        setCurrentTab={this.setCurrentTab}
        currentTab={currentTab}
        items={items}
        tabActive={tabActive}
      >
        {children}
      </TabKikoi>
    )
  }
}

export default TabKikoiContainer
