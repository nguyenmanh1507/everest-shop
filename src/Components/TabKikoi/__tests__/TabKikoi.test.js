// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { render, configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { trimStart, split } from 'lodash'

import TabKikoi from '../TabKikoiContainer'

configure({ adapter: new Adapter() })

test('TabKikoi render properly', () => {
  const component = renderer.create(renderTabKikoi())

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<TabKikoi />', () => {
  test('Should has tab item equal tab content', () => {
    const wrapper = render(renderTabKikoi())
    const totalTabContent = wrapper.find('[data-test^="tab-content"]').length
    const tabItem = wrapper.find('[data-test^="tab-item"]')
    expect(tabItem).toHaveLength(totalTabContent)
  })

  test('Should throw error if tab item number difference tab content number', () => {
    // TODO: Throw error if pass more content than items length and vice versa
  })

  test('Tab item has index equal default tab active should has active class', () => {
    const tabActive = 2
    const wrapper = render(renderTabKikoi(undefined, tabActive))
    const tabItem = wrapper.find(`[data-test="tab-item-${tabActive}"]`)

    expect(tabItem.hasClass('active')).toBeTruthy()
  })

  test('Should has only one tab item has active class at time', () => {
    const wrapper = mount(renderTabKikoi())
    const lastTabItem = wrapper.find('[data-test^="tab-item"]').last()
    const tabItem = wrapper.find('[data-test^="tab-item"].active')

    expect(tabItem).toHaveLength(1)

    // click last tab item
    lastTabItem.simulate('click')
    // find tab item has active class again
    const newTabItem = wrapper.find('[data-test^="tab-item"].active')
    expect(newTabItem).toHaveLength(1)
  })

  test('Should has only one tab content has active class at time', () => {
    const wrapper = render(renderTabKikoi())
    const tabItem = wrapper.find('[data-test^="tab-content"].active')

    expect(tabItem).toHaveLength(1)
  })

  test('Render tab item with active class when clicked', () => {
    const wrapper = mount(renderTabKikoi(undefined, 1))
    let tabItem = wrapper.find('[data-test^="tab-item"]').last()

    // make sure tab item not has active class
    expect(tabItem.hasClass('active')).toBeFalsy()
    tabItem.simulate('click')

    // select tab item after clicked
    tabItem = wrapper.find('[data-test^="tab-item"]').last()
    expect(tabItem.hasClass('active')).toBeTruthy()
  })

  test('Tab index has active class equal tab content index has active class', () => {
    const items = ['Description', 'Reviews', 'FAQ']
    const wrapper = mount(renderTabKikoi(items))
    // select Reviews tab
    let tabItem = wrapper.find('[data-test="tab-item-2"]')

    // click this tab
    tabItem.simulate('click')

    // select Reviews tab after clicked
    tabItem = wrapper.find('[data-test="tab-item-2"]')
    // select Reviews content
    const tabContent = wrapper.find('[data-test="tab-content-2"]')

    expect(tabItem.hasClass('active')).toBeTruthy()
    expect(tabContent.hasClass('active')).toBeTruthy()
  })
})

function renderTabKikoi(
  items = ['Description', 'Reviews', 'FAQ'],
  tabActive = 1
) {
  return (
    <TabKikoi items={items} tabActive={tabActive}>
      {items.map((item, index) => (
        <TabKikoi.Content key={`tab-content-test-${index + 1}`}>
          <div>{`${item} content go here`}</div>
        </TabKikoi.Content>
      ))}
    </TabKikoi>
  )
}
