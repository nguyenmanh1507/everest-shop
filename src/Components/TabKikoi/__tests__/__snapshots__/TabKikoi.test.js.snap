// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabKikoi render properly 1`] = `
<div
  className="tab"
>
  <div
    className="tab__nav tab__nav--center"
  >
    <button
      className="tab__nav-item active"
      data-item={1}
      data-test="tab-item-1"
      onClick={[Function]}
    >
      Description
    </button>
    <button
      className="tab__nav-item"
      data-item={2}
      data-test="tab-item-2"
      onClick={[Function]}
    >
      Reviews
    </button>
    <button
      className="tab__nav-item"
      data-item={3}
      data-test="tab-item-3"
      onClick={[Function]}
    >
      FAQ
    </button>
  </div>
  <div
    className="tab__body"
  >
    <div
      className="tab__content active"
      data-test="tab-content-1"
    >
      <div>
        Description content go here
      </div>
    </div>
    <div
      className="tab__content"
      data-test="tab-content-2"
    >
      <div>
        Reviews content go here
      </div>
    </div>
    <div
      className="tab__content"
      data-test="tab-content-3"
    >
      <div>
        FAQ content go here
      </div>
    </div>
  </div>
</div>
`;
